docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,446 @@
+
+## Proposal
+In this section we outline the proposed API changes; describe states, transitions, and sequence diagrams for the most important scenarios; and provide a kubeadm-based example.
+
+### Data model changes
+The Machine Spec and Status fields undergo some breaking changes which remove or add new fields. We introduce a Bootstrap type which encapsulates the integration point between a Machine and its path to become a Kubernetes Node.
+
+```go
+type MachineSpec struct
+```
+- **To remove**
+    - **ProviderSpec [optional]** _Superseded by InfrastructureRef_
+        - Type: `*runtime.RawExtension`
+        - Description: ProviderSpec details Provider-specific configuration to use during node creation.
+    - **Versions [optional]**  _Superseded by Version_
+        - Type: `MachineVersionInfo`
+        - Description: Versions of key software to use. This field is optional at cluster creation time, and omitting the field indicates that the cluster installation tool should select defaults for the user. These defaults may differ based on the cluster installer, but the tool should populate the values it uses when persisting Machine objects. A Machine spec missing this field at runtime is invalid.
+- **To add**
+    - **Bootstrap**
+        - Type: `Bootstrap`
+        - Description: Bootstrap is a reference to a local struct which encapsulates fields to configure the Machine’s bootstrapping mechanism.
+    - **InfrastructureRef [required]**
+        - Type: `corev1.TypedLocalObjectReference`
+        - Description: InfrastructureRef is a required reference to a Custom Resource Definition offered by an infrastructure provider.
+    - **Version [optional]**
+        - Type: `string`
+        - Description: Version defines the desired Kubernetes version. This field is meant to be optionally used by bootstrap providers.
+
+```go
+type MachineStatus struct
+```
+
+- **To remove**
+    - **ProviderStatus [optional]**
+        - Type: `*runtime.RawExtension`
+        - Description: ProviderStatus details a Provider-specific status. It is recommended that providers maintain their own versioned API types that should be serialized/deserialized from this field.
+    - **Versions [optional]**  _Superseded by Version_
+        - Type: `*MachineVersionInfo`
+        - Description: Versions specifies the current versions of software on the corresponding Node (if it exists). This is provided for a few reasons [...].
+    - **Phase [optional]**  _Superseded by Phase_
+        - Type: `*string`
+        - Description: Phase represents the current phase of machine actuation. E.g. Pending, Running, Terminating, Failed etc.
+- **To add**
+    - **Phase**
+        - Type: `MachinePhaseType`
+        - Description: Phase represents the current state of a Machine. This field is not meant to be used as a source of truth. Controllers working on Machines should always respect the documented transition conditions to determine the state.
```

Comment:
At one point we discussed removing the single "phase" field in favor of having a separate status field for each aspect we expect to be managed by a separate controller (see BootstrapPhase and InfrastructurePhase below). If we still want to do that, I can make those edits in a follow-up PR.

Reply:
You are right, top-level fields or conditions were discussed. Personally, I would like to consider the ["outlook" model](https://github.com/kubernetes/kubernetes/issues/7856#issuecomment-99667941). History has proven this could be a long conversation... :)

Reply:
I second @dhellmann proposal. Having conditions is more flexible and considering we are moving to implement v1alpha2 as a mean of learning, leaving room for extensions without force breaking changes is an interesting strategy.

Reply:
If we have individual "statuses" that we can represent as separate boolean values, +1 to that. I am opposed to an arbitrary array of strings for conditions, however.

Reply:
I'd definitely like to see a better approach to a string-based-Phase as a follow-up CAEP 👍

Reply:
@davidewatson I also like the outlook model here, would be interested in chatting about this for potential follow up.